### PR TITLE
Support `ENVELOPE DEBEZIUM UPSERT`

### PR DESCRIPTION
**Pull request by quodlibetor**
This adds upsert semantics support for Debezium sources. Fundamentally that means that we treat
upstream diffs as potentially nonsensical.

To do that, we track keys per-key values per-key, and only retract rows that we know that we have
seen.

There is one todo in the code that will be the immediate next work: with the code as it stands we
treat the keys as raw bytes, not decoding them, but since Avro is not self-describing it is possible
for the same bit pattern to have different values when decoded. We feel like this is unlikely enough
to be an issue for most users (considering especially that we don't really support schema evolution)
that it's probably safe to use as is.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/6210)

### CODE REVIEW

**Comment by quodlibetor** [comment]
@umanwizard this isn't ready for bikesheddy review but the core logic in `decode_values_inner` should match what we discussed. Feel free to wait to review until I make the code a bit less duplicated.

**Comment by quodlibetor** [comment]
Actually a particularly useful review would be of the added testdrive file, if you can think of extra tests to add to check edge cases that'd be useful.

File: src/dataflow/src/decode/mod.rs [hunk 1]
```diff
@@ -64,15 +60,18 @@ where
...
             &schema,
             debug_name.to_string(),
             stream.scope().index(),
             dedup_strat,
-        )
-    } else {
-        None
+        ),
+        SourceEnvelope::Debezium(_, DebeziumMode::Upsert) => {
+            error!("Upsert doesn't make sense for Avro OCFs");
```
Comment (umanwizard) [inline src/dataflow/src/decode/mod.rs:71]: you could say "because they don't have keys"

File: src/dataflow/src/decode/mod.rs [hunk 2]
```diff
@@ -292,6 +292,84 @@ where
...
+                } else {
+                    let mut packer = RowPacker::new();
+                    packer.push(Datum::Null);
+                    packer.push(after);
+                    Ok(packer.finish())
+                }
+            }
+            Entry::Occupied(mut occupied) => {
+                if occupied.get().iter().next() == Some(before) {
+                    if after.is_null() {
```
Comment (umanwizard) [inline src/dataflow/src/decode/mod.rs:342]: Don't we need an `else` branch here to set the value in the map to the new value?

Comment (quodlibetor) [reply src/dataflow/src/decode/mod.rs:342]: ah, yes, I'll add a test that checks for that too.

**Review by umanwizard — APPROVED** [review]
LGTM with a few caveats:

(1) actually parsing the key, instead of using raw bytes, should be a fast-follower.
(2) a fast-follower after _that_ should be a refactor of upserts in general to merge the different code paths, as discussed in https://gist.github.com/umanwizard/b7daf3fe6ae0f706b57aebbb064c1835 and elsewhere.

**Comment by quodlibetor** [comment]
> (1) actually parsing the key, instead of using raw bytes, should be a fast-follower.
> (2) a fast-follower after that should be a refactor of upserts in general to merge the different code paths, as discussed in https://gist.github.com/umanwizard/b7daf3fe6ae0f706b57aebbb064c1835 and elsewhere.

Yup, both of these are my planned next steps.
